src/core/geocms/geonode/qgsgeonoderequest.cpp

```diff
@@ -269,111 +269,106 @@ QList<QgsGeoNodeRequest::ServiceLayerDetail> QgsGeoNodeRequest::parseLayers( con
   const QJsonObject jsonObject = jsonDocument.object();
   const QVariantMap jsonVariantMap = jsonObject.toVariantMap();
   const QVariantList layerList = jsonVariantMap.value( QStringLiteral( "objects" ) ).toList();
-  qint16 majorVersion;
-  qint16 minorVersion;
-  if ( jsonVariantMap.contains( QStringLiteral( "geonode_version" ) ) )
+
+  QString wmsURLFormat, wfsURLFormat, xyzURLFormat;
+
+  for ( const QVariant &layer : qgis::as_const( layerList ) )
   {
-    QRegularExpression re( "((\\d+)(\\.\\d+))" );
-    QRegularExpressionMatch match = re.match( jsonVariantMap.value( QStringLiteral( "geonode_version" ) ).toString() );
-    if ( match.hasMatch() )
+    QgsGeoNodeRequest::ServiceLayerDetail layerStruct;
+    const QVariantMap layerMap = layer.toMap();
+    QVariantList layerLinks = layerMap.value( QStringLiteral( "links" ) ).toList();
+
+    if ( layerMap.value( QStringLiteral( "typename" ) ).toString().isEmpty() )
     {
-      const QStringList geonodeVersionSplit = match.captured( 0 ).split( '.' );
-      majorVersion = geonodeVersionSplit.at( 0 ).toInt();
-      minorVersion = geonodeVersionSplit.at( 1 ).toInt();
+      const QStringList splitURL = layerMap.value( QStringLiteral( "detail_url" ) ).toString().split( '/' );
+      layerStruct.typeName = splitURL.at( splitURL.length() - 1 );
     }
-    else
+    layerStruct.uuid = layerMap.value( QStringLiteral( "uuid" ) ).toString();
+    layerStruct.id = layerMap.value( QStringLiteral( "id" ) ).toString();
+    layerStruct.name = layerMap.value( QStringLiteral( "name" ) ).toString();
+    layerStruct.typeName = layerMap.value( QStringLiteral( "typename" ) ).toString();
+    layerStruct.title = layerMap.value( QStringLiteral( "title" ) ).toString();
+
+    layerStruct = parseOWSUrl( layerStruct, layerLinks );
+
+    if ( layerStruct.wmsURL.isEmpty() && layerStruct.wfsURL.isEmpty() && layerStruct.xyzURL.isEmpty() )
     {
-      return layers;
+      if ( wmsURLFormat.isEmpty() && wfsURLFormat.isEmpty() && xyzURLFormat.isEmpty() )
+      {
+        bool success = requestBlocking( QStringLiteral( "/api/layers/" ) + layerStruct.id );
+        if ( success )
+        {
+          const QJsonDocument resourceUriDocument = QJsonDocument::fromJson( this->lastResponse() );
+          const QJsonObject resourceUriObject = resourceUriDocument.object();
+          const QVariantMap resourceUriMap = resourceUriObject.toVariantMap();
+          QVariantList resourceUriLinks = resourceUriMap.value( QStringLiteral( "links" ) ).toList();
+          QgsGeoNodeRequest::ServiceLayerDetail tempLayerStruct;
+          tempLayerStruct = parseOWSUrl( tempLayerStruct, resourceUriLinks );
+
+          // Avoid iterating all the layers to get the service url. Instead, generate a string format once we found one service url
+          // for every service (wms, wfs, xyz). And then use the string format for the other layers since they are identical.
+          if ( tempLayerStruct.server == "qgis-server" )
```

Comment:
```suggestion
          if ( tempLayerStruct.server == QStringLiteral( "qgis-server" ) )
```

Reply:
ah yes of course, I must have missed some strings. thanks